rfcs/20240311-gather-scatter-batching-dims.md

```diff
@@ -0,0 +1,456 @@
+# [RFC] Add batching dims to `stablehlo.gather` and `stable.scatter` specification
+
+Status: Review<br/>
+Initial version: 03/11/2024<br/>
+Last updated: 03/11/2024<br/>
+Discussion thread: TBD
+
+## Overview
+
+This RFC proposes adding `operand_batching_dims` and
+`start_indices_batching_dims` attributes to `stablehlo.gather`.
+`operand_batching_dims` refers to the dimensions of the `operand` that are
+treated as batch. `start_indices_batching_dims` refers to the dimensions of the
+`start_indices` that are treated as batch. The corresponding dimension sizes
+must be equal. The semantics is equivalent to concatenating the outputs of the
+gather with each slices of `operand` and `start_indices`.
+
+Similarly, this RFC proposes adding `input_batching_dims` and
+`scatter_indices_batching_dims` attributes to `stablehlo.scatter`.
+`input_batching_dims` refers to the dimensions of each tensor in `inputs` that
+are treated as batch. `scatter_indices_batching_dims` refers to the dimensions
+of the `scatter_indices` that are treated as batch.
+
+## Motivation
+
+StableHLO gather and scatter ops currently have no way of specifying batch
+dimensions that correspond between all operands and result, only the indices
+tensor can have implicit batch dimensions that only exist in the
+result/update tensor (`batch_dims` in specification).
+
+This is important when the user wants a batched/vectorized version of
+gather/scatter across all operands and result (e.g., when using [jax.vmap](https://jax.readthedocs.io/en/latest/_autosummary/jax.vmap.html)).
+
+The current workaround is to use the implicit batch dimensions in the indices
+tensor (all dimensions but `index_vector_dim`) with `stablehlo.concatenate` and
+`stablehlo.iota` (and other ops like `stablehlo.clamp`) to mimic batch
+dimensions in the operand.
+
+This isn't ideal as it hides the fact that those are batch dimensions that
+correspond between the operands and result tensors. This information is crucial
+when doing sharding propagation between the operands and result tensors of the
+gather/scatter op, where partitioning batch dimensions across tensors is
+trivial, as they can be sharded in the same way with no communication needed.
+The above workaround requires pattern matching to identify those batch
+dimensions, which can be error prone and hard to maintain.
+
+This proposal is inspired by `lhs_batching_dims` and `rhs_batching_dims` of
+`stablehlo.dot_general`, which serve the same purpose.
+
+## Compatibility
```

Comment:
How would this work for HLO<->StableHLO interop? Is this just a 1:1 update on the respective HLO ops?

Reply:
Yes, the plan is to also add these attributes to the respective HLO ops, but this might take more time, in which case when going from StableHLO to HLO, we will remove the batching dims using the workaround with iota. Going from HLO to StableHLO will remain the same.

Does that make sense?

Reply:
Thanks, yes I think so. In that case though, would it be roundtrippable or would we lose this attribute when going from StableHLO -> HLO -> StableHLO?

Reply:
We would, unless we implement some canonicalization pattern that pattern matches on iota and adds the new batch dimensions, but since this is only temporary I would just push on the XLA change so roundtripping will work properly.

Reply:
Ack thanks!